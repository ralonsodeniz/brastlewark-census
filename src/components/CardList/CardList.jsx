import React, { useCallback } from 'react';
import { useDispatch } from 'react-redux';
import PropTypes from 'prop-types';

import { openModal } from '../../redux/actions/modalActions';

import Card from '../Card/Card';
import CardSkeleton from '../CardSkeleton/CardSkeleton';
import Spinner from '../Spinner/Spinner';

import { CardListContainer, CardListSpinnerContainer } from './CardList.style';

const CardList = ({ items, loading }) => {
  const dispatch = useDispatch();

  const handleCardClick = useCallback(
    item => {
      dispatch(
        openModal({
          modalType: 'USER_DETAIL',
          modalProps: item,
        })
      );
    },
    [dispatch]
  );

  const CardsMarkUp = !loading
    ? items.map(item => {
        const { name, thumbnail, id } = item;
        return (
          <Card key={id} imageUrl={thumbnail} name={name} onClick={() => handleCardClick(item)} />
        );
      })
    : Array.from({ length: 4 }).map((card, cardIndex) => <CardSkeleton key={cardIndex} />);

  return (
    <CardListContainer>
      {loading && (
        <CardListSpinnerContainer>
          <Spinner />
        </CardListSpinnerContainer>
      )}
      {CardsMarkUp}
    </CardListContainer>
  );
};

CardList.defaultProps = {
  loading: false,
};

CardList.propTypes = {
  items: PropTypes.arrayOf(PropTypes.object).isRequired,
  loading: PropTypes.bool,
};

export default CardList;
