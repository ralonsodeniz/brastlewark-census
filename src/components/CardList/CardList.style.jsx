import styled, { css } from 'styled-components';

import mediaQueryHelper from '../../style/media-queries';

export default styled.div`
  display: grid;
  grid-template-columns: repeat(auto-fill, 250px);
  /* grid-template-rows: repeat(auto-fill, 250px); */
  padding: 1rem 0;
  justify-content: center;
  grid-gap: 20px;
  width: 100%;
  height: min-content;
  max-height: 61vh;
  background: transparent;
  margin-top: 30px;
  overflow-x: hidden;
  overflow-y: auto;
  box-shadow: inset 0 0 6px rgba(0, 0, 0, 0.3);
  -webkit-box-shadow: inset 0 0 6px rgba(0, 0, 0, 0.3);

  ${mediaQueryHelper(
    'tab-port',
    css`
      padding: 0.5rem 0;
      grid-template-columns: repeat(auto-fill, 200px);
    `
  )};

  ${mediaQueryHelper(
    'phone',
    css`
      padding: 0.5rem 0;
      grid-template-columns: repeat(auto-fill, 125px);
    `
  )};

  /* overflow */
  ::-webkit-scrollbar {
    height: 12px;
    display: inline-block;
  }

  ::-webkit-scrollbar-track {
    box-shadow: inset 0 0 6px rgba(0, 0, 0, 0.3);
    -webkit-box-shadow: inset 0 0 6px rgba(0, 0, 0, 0.3);
    -webkit-border-radius: 10px;
    border-radius: 10px;
  }

  ::-webkit-scrollbar-thumb {
    -webkit-border-radius: 10px;
    border-radius: 10px;
    background: rgba(0, 0, 0, 0.4);
    box-shadow: inset 0 0 6px rgba(0, 0, 0, 0.5);
    -webkit-box-shadow: inset 0 0 6px rgba(0, 0, 0, 0.5);
  }

  ${mediaQueryHelper(
    'phone',
    css`
      margin-top: 20px;
      max-height: 50vh;
    `
  )};

  ${mediaQueryHelper(
    'phone-xs',
    css`
      margin-top: 20px;
      max-height: 43vh;
      grid-gap: 13px;
    `
  )};
`;
