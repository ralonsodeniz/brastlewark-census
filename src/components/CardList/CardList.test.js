/* eslint-disable prefer-const */
import React from 'react';
import * as Redux from 'react-redux';
import { shallow } from 'enzyme';
import configureStore from 'redux-mock-store';

import CardList from './CardList';

describe('CardList Component tests', () => {
  let wrapper;
  let store;
  let useEffect;

  const INITIAL_STATE = {
    modal: {
      showModal: false,
      modalData: {
        modalType: '',
        modalProps: {},
      },
    },
    data: {
      loadingData: false,
      gnomeNameSearchField: '',
      professionSearchField: '',
      gnomesData: [],
      error: '',
    },
  };

  const mockUseEffect = () => {
    useEffect.mockImplementationOnce(f => f());
  };

  store = configureStore([])(INITIAL_STATE);

  useEffect = jest.spyOn(React, 'useEffect');

  jest.spyOn(Redux, 'useDispatch').mockImplementation(() => store.dispatch);

  test('expect CardList to match shallow snapshot', () => {
    const mockProps = {
      items: [],
      loading: false,
    };

    wrapper = shallow(<CardList {...mockProps} store={store} />);

    expect(wrapper).toMatchSnapshot();
  });

  describe('CardList rendering Skeletons', () => {
    test('expect CardList to render 4 CardSkeletons when items is empty array', () => {
      const mockProps = {
        items: [],
        loading: false,
      };

      wrapper = shallow(<CardList {...mockProps} store={store} />);

      const cardSkeletons = wrapper.find('CardSkeleton');
      expect(cardSkeletons).toHaveLength(4);
    });

    test('expect CardList to render 4 CardSkeletons when height is not set', () => {
      const mockProps = {
        items: [
          {
            name: 'John',
            thumbnail: 'placeholder.com',
            id: 1,
          },
        ],
        loading: false,
      };
      wrapper = shallow(<CardList {...mockProps} store={store} />);

      const cardSkeletons = wrapper.find('CardSkeleton');
      expect(cardSkeletons).toHaveLength(4);
    });
  });

  describe('CardList rendering Cards', () => {
    test('expect CardList to have one Card Component with height 125', () => {
      global.innerWidth = 500;
      global.dispatchEvent(new Event('resize'));

      mockUseEffect();
      mockUseEffect();

      const mockProps = {
        items: [
          {
            name: 'John',
            thumbnail: 'placeholder.com',
            id: 1,
          },
        ],
        loading: false,
      };
      wrapper = shallow(<CardList {...mockProps} store={store} />);

      const card = wrapper.find('Card');
      expect(card).toHaveLength(1);
      expect(card.prop('height')).toEqual(125);
    });

    test('expect CardList to have one Card Component with height 250', () => {
      global.innerWidth = 1200;
      global.dispatchEvent(new Event('resize'));

      mockUseEffect();
      mockUseEffect();

      const mockProps = {
        items: [
          {
            name: 'John',
            thumbnail: 'placeholder.com',
            id: 1,
          },
        ],
        loading: false,
      };
      wrapper = shallow(<CardList {...mockProps} store={store} />);

      const card = wrapper.find('Card');
      expect(card).toHaveLength(1);
      expect(card.prop('height')).toEqual(250);
    });

    test('expect CardList to have one Card Component with height 200', () => {
      global.innerWidth = 800;
      global.dispatchEvent(new Event('resize'));

      mockUseEffect();
      mockUseEffect();

      const mockProps = {
        items: [
          {
            name: 'John',
            thumbnail: 'placeholder.com',
            id: 1,
          },
        ],
        loading: false,
      };
      wrapper = shallow(<CardList {...mockProps} store={store} />);

      const card = wrapper.find('Card');
      expect(card).toHaveLength(1);
      expect(card.prop('height')).toEqual(200);
    });

    test('expect CardList to dispatch proper action on click', () => {
      global.innerWidth = 500;
      global.dispatchEvent(new Event('resize'));

      mockUseEffect();
      mockUseEffect();

      const mockProps = {
        items: [
          {
            name: 'John',
            thumbnail: 'placeholder.com',
            id: 1,
          },
        ],
        loading: false,
      };
      wrapper = shallow(<CardList {...mockProps} store={store} />);

      const card = wrapper.find('Card');
      card.simulate('click');
      const actions = store.getActions();
      expect(actions).toEqual([
        {
          payload: {
            modalProps: { id: 1, name: 'John', thumbnail: 'placeholder.com' },
            modalType: 'USER_DETAIL',
          },
          type: 'OPEN_MODAL',
        },
      ]);
    });
  });
});
