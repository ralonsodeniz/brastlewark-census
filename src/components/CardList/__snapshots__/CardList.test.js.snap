// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardList Component tests expect CardList to match shallow snapshot 1`] = `
<styled.div>
  <CardSkeleton
    key="0"
  />
  <CardSkeleton
    key="1"
  />
  <CardSkeleton
    key="2"
  />
  <CardSkeleton
    key="3"
  />
</styled.div>
`;
